### ISSUE TITLE
Lua Variables Indexed section starts at 1 instead of 0

### ISSUE THREAD

**Issue by justarandomgeek** [microsoft/debug-adapter-protocol#104]
Lua tables are both array and hashtable, but the array part is integer indexes starting at 1, and 0 is just another hashtable key in the hashtable part. This requires me to fudge the boundaries of paged variables requests, where DAP and tools currently assume starting at 0. I currently [report indexedVariables = length+1](https://github.com/justarandomgeek/vscode-factoriomod-debug/blob/master/mod/variables.lua#L280) of an array and just [omit an entry for 0](https://github.com/justarandomgeek/vscode-factoriomod-debug/blob/master/mod/variables.lua#L459) in the Indexed section, but this causes some annoyances in the UI, which still indicates that the first page starts as `0..99`. It would be great if I could somehow indicate that Indexed variables in my environment start from 1 and have the dev tool adjust its UI and paged requests to me accordingly.

**Comment by weinand** [comment]
This feature request makes a lot of sense.
We could either have a capability for this ('arrays index starts at 1') which makes the client (frontend) use ranges from 1..100 , or we could add an attribute to some response that allows to override the string "0..99" that the client shows (this would still need the "fudging" you are doing today).

**Comment by justarandomgeek** [comment]
A capability was my initial thought as well, but then I looked at the existing ones and it felt out of place - All the other capabilities are "I can do X if you want it" and if the client ignores it nothing happens, while this is more of "my X is Y" and ignoring it would cause the current behavior which requires special treatment to work correctly. It's also possible that some API from the environment in which Lua is embedded might present objects which *are* zero-indexed (because they come from the rest of the world unadjusted), so maybe it should be attached to the individual objects (VariablePresentationHint?) instead of the whole session? (I don't think I have any specific examples of this because Factorio is very consistent about adjusting indexes at the boundary, but I'm sure other environments are less consistent.) It does seem like I'll probably have to handle it being potentially ignored in any case, since the whole world is zero based!

As for what should be done with the label/request windows when presented with this capability or hint, it seems like there's two options really:
A) shift all the windows up one, so all the groups are the same size still: 1..100,101..200,201..300,...
B) just skip 0 and keep the windows grouped by leading digits: 1..99,100..199,200..299,...

This really seems like a toss-up to me, and can/should probably even be left up to the client to decide (the spec seems to leave the size of the windows up to the client already), but if I had to pick, I'd probably take B. Being grouped by leading digit seems like it makes it slightly easier to find an item in a large collection (no surprises where x00 is in the previous group). It's also essentially what I'm coercing the existing requests to by skipping the 0, and seems like it should be a fairly minimal change for clients.

**Comment by connor4312** [comment]
> or we could add an attribute to some response that allows to override the string "0..99" that the client shows (this would still need the "fudging" you are doing today).

I don't think this is a good solution, since it's not just the preview string, but also the indexes once that's expanded (at least for vscode which shows indexes)

<img width="236" alt="image" src="https://user-images.githubusercontent.com/2230985/202252161-b59ec100-a243-450c-b6db-29827277f177.png">

I think having something in the `capabilities` event, but potentially not in the "Capabilities" object, makes sense. Thus the event would look something like:

```ts
interface ICapabilitiesEvent {
  capabilities: { /* ... */ };
  
  /**
   * If true, when indexed variables are presented in the UI, the first indexed
   * variable should be displayed at index 1 rather than index 0.
   * 
   * Note that this property does not change any protocol semantics. For example,
   * the `start` property in the `variables` should be set to 0 to include
   * the first indexed variable regardless of the value of this property.
   */
  indexedVariablesStartAt1?: boolean;
}
```
